server/src/main/java/io/crate/types/NumericType.java

```diff
@@ -123,21 +124,25 @@ public BigDecimal implicitCast(Object value) throws IllegalArgumentException, Cl
             return null;
         }
 
-        var mathContext = mathContext();
         BigDecimal bd;
         if (value instanceof BigDecimal bigDecimal) {
-            bd = bigDecimal.round(mathContext);
+            bd = bigDecimal;
         } else if (value instanceof String || value instanceof Float || value instanceof Double) {
-            bd = new BigDecimal(value.toString(), mathContext);
+            bd = new BigDecimal(value.toString());
         } else if (value instanceof Number number) {
-            bd = new BigDecimal(BigInteger.valueOf(number.longValue()), mathContext);
+            bd = new BigDecimal(BigInteger.valueOf(number.longValue()));
         } else {
-            throw new ClassCastException("Can't cast '" + value + "' to " + getName());
+            throw new ClassCastException("Cannot cast '" + value + "' to " + getName());
         }
         if (scale == null) {
             return bd;
         }
-        return bd.setScale(scale, mathContext.getRoundingMode());
+        int sourceNumIntegralDigits = bd.precision() - bd.scale();
+        int targetNumIntegralDigits = precision - scale; // since scale != null, precision is also != null
+        if (sourceNumIntegralDigits - targetNumIntegralDigits > 0) {
+            throw new ClassCastException("Cannot cast '" + value + "' to " + this + " as it looses precision");
+        }
+        return bd.setScale(scale, RoundingMode.HALF_UP);
```

Comment:
I wonder if we should stay consistent and keep using the `mathcontext.getRoundingMode()` instead even that it currently will always be `RoundingMode.HALF_UP`.

Reply:
Sorry, thx, we should keep it yes, just forgot to revert this change while testing something.